### ISSUE TITLE
[Feature Request] Sideloading (mod) files via http/https

### ISSUE THREAD

**Issue by ger-pakon** [BeamMP/BeamMP-Server#9]
Hey,

would be great, if there is a setting (maybe a url) locating to a webserver, where the mods can be downloaded.
The folderstructure must be the same as in the game, may makes it more easy. The Client will download the mods from this url instead from the BeamMP-Server.

Cheers,
Pakon

**Comment by Starystars67** [comment]
What would be the intended gain / benefit of doing this?
I am not putting it/you down, just trying to get an idea as to what you are trying to gain by it.

**Comment by ger-pakon** [comment]
Hey @Starystars67
I see some benefits:
- reducing the "load" of the server (actually its good at all)
- reduces bandwith on the beammp server
- use a CDN or another webserver to increase the speed of the download
  - a own caching - reverse proxy would be possible

Thanks,
Pakon

**Comment by lionkor** [comment]
I'd like to add that while this is a nice idea, the biggest benefit I see would be that you can host mods from a different server than the BeamMP-Server. Considering the tiny system footprint of a server (even when maxed out with 15+ players), I don't quite see why the server wouldn't live on the same machine as the mods. 
I do agree that we should allow the Server Lua to pass FTP/SFTP commands in order to self-update or auto-download stuff. I wouldn't mind hearing other ideas for why/how/if this should be done, feel free to discuss on this issue.

EDIT: I had not considered CDNs

**Comment by Starystars67** [comment]
@lionkor 
> Server Lua to pass FTP/SFTP 

This is a bad idea in my opinion as it could be exploited if someone uses a server resource that had malicious intent. HTTP/s post / get requests would be better as it helps to reduce the chance of uploading details / data. I dont know it just feels off to have FTP access for something that shouldn't need it.

**Comment by Ae3NerdGod** [comment]
I agree with not using ftp, but I've started using what I'm calling modpacks on a private server because they make the game load much quicker (~60 mods, 10gb loads in seconds as if the server is vanilla)

 but beammp-server often fails to serve these files to my users via its own methods. 

Being able to serve it from a sourceforge link for example would be greatly helpful.

**Comment by Starystars67** [comment]
Maybe we could introduce a system where you could position NGINX or similar serving systems. The question is is the issue with your upload or something else?

We are working on fixing the downloading process that was partially broken in the latest update.

**Comment by Ae3NerdGod** [comment]
Ah, I wasn't aware there was an actual issue with the server, I just assumed it was my connection actually. I've been uploading the packs to google drive with my buddies fiber from that assumption and did no further investigation. I can fire up nginx and see if they download them directly from my server to test later if/when we get on

**Comment by Starystars67** [comment]
No worries @Ae3NerdGod , Sorry for the confusion there. In short the error is that for some reason the launcher is being buggy with the download which messes with the UI in game and in turn everything seems slow and janky. The good news is that we are working on it. Something to note is that the BeamMP server actually does try to serve the files at maximum speed for you to download at and if you were to try and test its speed it should be where you would expect it to be (Allowing for the slowest point in the network be it the server or the client).

**Comment by lionkor** [comment]
The builtin HTTPS server (#69, #68) could serve the local files, this would simplify the download mechanism to a HTTP/S GET on the client-side, and a simple file serve on the server side. It could also be ratelimited fairly well.

**Comment by lionkor** [comment]
@jimkoen @Starystars67

**Comment by lionkor** [comment]
This is planned for the rust-server branch.
